core/src/main/scala/chisel3/experimental/hierarchy/Hierarchy.scala

```diff
@@ -21,8 +21,11 @@ sealed trait Hierarchy[+A] {
     case Clone(i: IsClone[A]) => i.getProto
   }
 
-  /** Updated by calls to [[_lookup]], to avoid recloning returned Data's */
-  private[chisel3] val cache = HashMap[Data, Data]()
+  /** Updated by calls to [[_lookup]], to avoid recloning returned values */
+  private val cache = HashMap[Any, Any]()
```

Comment:
I'm not sure if I can construct the case where it will matter, but my intuition is still that this should be an `IdentityHashMap`